src/Features/Core/Portable/ConvertTupleToStruct/AbstractConvertTupleToStructCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,865 @@
+
+            var updatedSolution = await ApplyChangesAsync(
+                document, documentToEditorMap, cancellationToken).ConfigureAwait(false);
+
+            return updatedSolution;
+        }
+
+        private async Task ReplaceExpressionAndTypesInScopeAsync(
+            Dictionary<Document, SyntaxEditor> documentToEditorMap, 
+            ImmutableArray<DocumentToUpdate> documentsToUpdate,
+            SyntaxNode tupleExprOrTypeNode, INamedTypeSymbol tupleType, 
+            string structName, ImmutableArray<ITypeParameterSymbol> typeParameters,
+            INamespaceSymbol containingNamespace, CancellationToken cancellationToken)
+        {
+            // Process the documents one project at a time.
+            foreach (var group in documentsToUpdate.GroupBy(d => d.Document.Project))
+            {
+                // grab the compilation and keep it around as long as we're processing
+                // the project so we don't clean things up in the middle.
+                var project = group.Key;
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+                var generator = project.LanguageServices.GetService<SyntaxGenerator>();
+
+                // Get the fully qualified name for the new type we're creating.  We'll use this
+                // at replacement points so that we can find the right type even if we're in a 
+                // different namespace.
+
+                // If the struct is being injected into the global namespace, then reference it with
+                // "global::NewStruct",  Otherwise, get the full name to the namespace, and append
+                // the NewStruct name to it.
+                var structNameNode = CreateStructNameNode(
+                    generator, structName, typeParameters, addRenameAnnotation: false);
+
+                var fullTypeName = containingNamespace.IsGlobalNamespace
+                    ? (TNameSyntax)generator.GlobalAliasedName(structNameNode)
+                    : (TNameSyntax)generator.QualifiedName(generator.NameExpression(containingNamespace), structNameNode);
+
+                fullTypeName = fullTypeName.WithAdditionalAnnotations(Simplifier.Annotation)
+                                           .WithAdditionalAnnotations(DoNotAllowVarAnnotation.Annotation);
+
+                foreach (var documentToUpdate in group)
+                {
+                    var document = documentToUpdate.Document;
+                    var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                    // If we were given specific nodes to update, only update those.  Otherwise
+                    // updated everything from the root down.
+                    var nodesToUpdate = documentToUpdate.NodesToUpdate.IsDefaultOrEmpty
```

Comment:
Should this just be `IsDefault`? Feels like the empty case means "we were given specific nodes".

Reply:
Updated and added assert.